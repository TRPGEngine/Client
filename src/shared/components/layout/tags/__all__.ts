import { registerTag } from '../tags';
import { TagColShared } from './Col/shared';
import { TagBaseAttrShared } from './BaseAttr/shared';
import { TagInputEdit } from './Input/edit';
import { TagUseShared } from './Use/shared';
import { TagDefineShared } from './Define/shared';
import { TagTextAreaEdit } from './TextArea/edit';
import { TagInputDetail } from './Input/detail';
import { TagSelectEdit } from './Select/edit';
import { TagTemplateShared } from './Template/shared';
import { TagBaseInfoEdit } from './BaseInfo/edit';
import { TagBaseInfoDetail } from './BaseInfo/detail';
import { TagRowShared } from './Row/shared';
import { TagTabsShared } from './Tabs/shared';
import { TagBarShared } from './Bar/shared';
import { TagHiddenShared } from './Hidden/shared';
import { TagVarShared } from './Var/shared';
import { TagFieldSetShared } from './FieldSet/shared';
import { TagRollBtnEdit } from './RollBtn/edit';
import { TagRollBtnDetail } from './RollBtn/detail';
import { TagComputedShared } from './Computed/shared';
import { TagTipShared } from './Tip/shared';
import { TagRadarShared } from './Radar/shared';
import { TagFunctionShared } from './Function/shared';
import { TagInputNumberEdit } from './InputNumber/edit';
import { TagStyledShared } from './Styled/shared';
import { TagScriptShared } from './Script/shared';
import { TagForEachShared } from './ForEach/shared';
import { TagStaticShared } from './Static/shared';
import { TagDataTableShared } from './DataTable/shared';
import { TagDividerShared } from './Divider/shared';
import { TagCheckboxDetail } from './Checkbox/detail';
import { TagCheckboxEdit } from './Checkbox/edit';
import { TagCustomListShared } from './CustomList/shared';
import { TagCustomListEdit } from './CustomList/edit';
import { TagCurrMaxEdit } from './CurrMax/edit';
import { TagCurrMaxDetail } from './CurrMax/detail';
import { TagSpaceShared } from './Space/shared';
import { TagButtonShared } from './Button/shared';

// 展示数据组件
registerTag('Template', TagTemplateShared);

registerTag('Row', TagRowShared);

registerTag('Col', TagColShared);

registerTag('BaseInfo', TagBaseInfoDetail, TagBaseInfoEdit);

registerTag('Tabs', TagTabsShared);

registerTag('Bar', TagBarShared);

registerTag('Hidden', TagHiddenShared);

registerTag('FieldSet', TagFieldSetShared);

registerTag('Tip', TagTipShared);

registerTag('DataTable', TagDataTableShared);

registerTag('Radar', TagRadarShared);

registerTag('Styled', TagStyledShared);

registerTag('Divider', TagDividerShared);

registerTag('Space', TagSpaceShared);

// 数据录入组件
registerTag('BaseAttr', TagBaseAttrShared);

registerTag('Input', TagInputDetail, TagInputEdit);

registerTag('InputNumber', TagInputDetail, TagInputNumberEdit);

registerTag('TextArea', TagInputDetail, TagTextAreaEdit);

registerTag('Select', TagInputDetail, TagSelectEdit);

registerTag('Checkbox', TagCheckboxDetail, TagCheckboxEdit);

registerTag('RollBtn', TagRollBtnDetail, TagRollBtnEdit);

registerTag('CurrMax', TagCurrMaxDetail, TagCurrMaxEdit);

registerTag('Button', TagButtonShared, TagButtonShared);

// 高级操作组件
registerTag('Use', TagUseShared);

registerTag('Define', TagDefineShared);

registerTag('Var', TagVarShared);

registerTag('Static', TagStaticShared);

registerTag('Computed', TagComputedShared);

registerTag('Function', TagFunctionShared);

registerTag('Script', TagScriptShared);

registerTag('ForEach', TagForEachShared);

registerTag('CustomList', TagCustomListShared, TagCustomListEdit);
