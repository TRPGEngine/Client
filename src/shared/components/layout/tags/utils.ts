import { Fragment } from 'react';
import _get from 'lodash/get';
import _set from 'lodash/set';
import _isNil from 'lodash/isNil';
import _clone from 'lodash/clone';
import _toPairs from 'lodash/toPairs';
import _fromPairs from 'lodash/fromPairs';
import type {
  XMLBuilderState,
  XMLBuilderContext,
} from '@shared/components/layout/XMLBuilder';
import {
  StateDataType,
  StateActionType,
} from '@shared/components/layout/types';
import type { XMLElement } from '../parser/xml-parser';

export type OperationDataType = {
  scope: string;
  field: string;
};

/**
 * 获取需要操作的变量的作用域与操作的变量名
 * 作用域默认为data
 * 用于独立一个path的第一段
 * example:
 * - 'a' => {scope: 'data', field: 'a'}
 * - 'a.b' => {scope: 'a', field: 'b'}
 * - 'a.b.c' => {scope: 'a', field: 'b.c'}
 *
 * @param str 操作参数字符串
 */
export const getOperationData = (str: string): OperationDataType => {
  const [scope, ...fields] = str.split('.');

  if (fields.length > 0) {
    // 如果为abc.def
    return {
      scope,
      field: fields.join('.'),
    };
  } else {
    // 如果为abc
    return {
      scope: 'data',
      field: scope,
    };
  }
};

/**
 * 尝试将文本转化为数字
 * 用于表单数据转换
 * @param str 要转换的文本
 */
export const tryToNumber = (str: string): number | string => {
  if (typeof str === 'string' && str === '') {
    // 空字符串直接返回空字符串
    return '';
  }

  if (typeof str === 'string' && (str.length > 15 || str.startsWith(' '))) {
    // 大于15位的数字视为字符串
    // 输入空格视为字符串
    return str;
  }

  const num = Number(str);
  return !isNaN(num) ? num : str;
};

/**
 * 标准化标签名。返回字符串或空节点
 * @param tagName 标签名
 */
export const normalizeTagName = (tagName: string): string | typeof Fragment => {
  // 如果是首字母开头。视为没有做定义的内置操作。改为t-xxx的格式防止抛出命名警告
  if (typeof tagName === 'string' && /[A-Z]/.test(tagName[0])) {
    tagName = 't-' + tagName.toLowerCase();
  }

  // 如果是空字符串或者undefined。使用react的Fragment
  if (!tagName) {
    return Fragment;
  }

  return tagName;
};

/**
 * 获取要实际操作变量名的路径
 *
 * 返回要获取变量在data数据中的位置
 */
export const getNamePath = (name: string) => {
  const { scope, field } = getOperationData(name);
  let path = field;
  if (scope !== 'data') {
    path = [scope, field].join('.');
  }

  return path;
};

/**
 * 根据上下文获取指定状态的数据
 */
export const getStateValue = (
  context: XMLBuilderContext,
  bindingName: string
): StateDataType => {
  const path = getNamePath(bindingName);

  let ret = _get(context.state.data, path);
  if (_isNil(ret)) {
    // 如果没有从data中取到。可能path是有前缀的路径。尝试从上一级获取
    ret = _get(context.state, path);
  }

  // NOTICE: 不知道为什么要不允许获取object
  // 直接返回匹配的数据，如果有问题再改回来
  // if (typeof ret === 'object') {
  //   return '';
  // }

  return ret as any;
};

/**
 * 设置状态的值
 */
export const setStateValue = (
  context: XMLBuilderContext,
  bindingName: string,
  value: StateDataType
): void => {
  const { dispatch } = context;
  const { scope, field } = getOperationData(bindingName);

  dispatch({
    type: StateActionType.UpdateData,
    payload: {
      scope,
      field,
      value,
    },
  });
};
/**
 * 将属性中额私有参数提取出来(以下划线开头_的成员变量)
 * 返回一个新的对象
 */
export const removePrivateProps = <T extends object>(props: T): Partial<T> => {
  return _fromPairs(
    _toPairs(props).filter(([key, value]) => !key.startsWith('_'))
  ) as T;
};

/**
 * 处理xml读取到的多行文本
 */
export const parseMultilineText = (text: string) => {
  // 支持\n的渲染 拿到的换行符为\\n
  if (_isNil(text) || typeof text !== 'string') {
    return '';
  }

  return text.replace(new RegExp('\\\\n', 'g'), '\n');
};

/**
 * 获取元素子节点的文本
 */
export const getChildrenText = (el: XMLElement): string => {
  const childrenEl = el.elements!;

  return childrenEl
    .map<string>((el) => {
      if (el.type === 'text') {
        return String(el.text);
      } else if (el.type === 'element') {
        return getChildrenText(el);
      } else {
        return '';
      }
    })
    .reduce((prev, curr) => {
      return prev + curr;
    }, '');
};
