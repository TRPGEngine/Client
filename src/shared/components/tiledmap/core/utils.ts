import type { Size, Position, Rect } from './types';

/**
 * 计算距离数最近的网格点
 * @param num 真实位置
 * @param gridSizeNum 网格大小
 */
function calcPx2GridMagnetNum(num: number, gridSizeNum: number): number {
  return Math.round(num / gridSizeNum);
}

/**
 * 计算距离数最左上角的点
 * @param num
 * @param gridSizeNum
 */
function calcPx2GridFloorNum(num: number, gridSizeNum: number): number {
  return Math.floor(num / gridSizeNum);
}

/**
 * 像素位置转网格位置
 */
export function px2gridPos(
  pos: Position,
  gridSize: Size,
  isMagnet: boolean = false
): Position {
  const calcFn = isMagnet ? calcPx2GridMagnetNum : calcPx2GridFloorNum;

  return {
    x: calcFn(pos.x, gridSize.width),
    y: calcFn(pos.y, gridSize.height),
  };
}

/**
 * 标准化Rect
 */
export function normalizeRect(rect: Rect): Rect {
  return {
    x1: Math.min(rect.x1, rect.x2),
    y1: Math.min(rect.y1, rect.y2),
    x2: Math.max(rect.x1, rect.x2),
    y2: Math.max(rect.y1, rect.y2),
  };
}

/**
 * 判断rect1是否在rect2中
 */
export function isRectInsideRect(rect1: Rect, rect2: Rect) {
  rect1 = normalizeRect(rect1);
  rect2 = normalizeRect(rect2);

  return (
    rect1.x1 >= rect2.x1 &&
    rect1.x2 <= rect2.x2 &&
    rect1.y1 >= rect2.y1 &&
    rect1.y2 <= rect2.y2
  );
}

/**
 * 将pxRect转化成gridRect
 * 算法是网里缩
 */
export function pxRect2GridRect(rect: Rect, gridSize: Size): Rect {
  rect = normalizeRect(rect);

  return {
    x1: Math.ceil(rect.x1 / gridSize.width),
    y1: Math.ceil(rect.y1 / gridSize.height),
    x2: Math.floor(rect.x2 / gridSize.width),
    y2: Math.floor(rect.y2 / gridSize.height),
  };
}

/**
 * 判断一个点是否在一个矩形空间内
 */
export function isPositionInsizeRect(pos: Position, rect: Rect): boolean {
  rect = normalizeRect(rect);

  return (
    pos.x <= rect.x2 && pos.x >= rect.x1 && pos.y <= rect.y2 && pos.y >= rect.y1
  );
}

/**
 * 根据宽高和中心点返回一个矩形
 */
export function getRectWithSize(pos: Position, size: Size): Rect {
  return {
    x1: pos.x - size.width / 2,
    y1: pos.y - size.height / 2,
    x2: pos.x + size.width / 2,
    y2: pos.y + size.height / 2,
  };
}
