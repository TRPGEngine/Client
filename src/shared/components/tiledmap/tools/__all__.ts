import type { Toolbox } from '../core/toolbox';
import { TiledMapToolMove } from './move';
import { TiledMapToolSelect } from './select';

export function regAllTool(toolbox: Toolbox) {
  toolbox.regTool(new TiledMapToolSelect());
  toolbox.regTool(new TiledMapToolMove());

  toolbox.setCurrentTool('select');
}
