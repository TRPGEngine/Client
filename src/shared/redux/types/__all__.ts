import type { AnyAction, Action, Store, Middleware, Reducer } from 'redux';
import type { ActorState } from './actor';
import type { CacheState } from './cache';
import type { ChatState } from './chat';
import type { GroupState } from './group';
import type { NoteState } from './note';
import type { SettingsState } from './settings';
import type { UIState } from './ui';
import type { UserState } from './user';
import {
  createAsyncThunk,
  AsyncThunkPayloadCreator,
  AsyncThunk,
} from '@reduxjs/toolkit';

interface AllState {
  actor: ActorState;
  cache: CacheState;
  chat: ChatState;
  group: GroupState;
  note: NoteState;
  settings: SettingsState;
  ui: UIState;
  user: UserState;
  [other: string]: any;
}

// 用于state声明
export type TRPGState = AllState;

export type TRPGStore = Store<TRPGState, any>;

export type TRPGReducer = Reducer<TRPGState, any>;

// fork from redux-thunk declaration
interface ThunkDispatch<S, E, A extends Action> {
  <T extends A>(action: T): T;
  <R>(asyncAction: TRPGAction): R;
}

type ThunkAction<R, S, E, A extends Action> = (
  dispatch: ThunkDispatch<S, E, A>,
  getState: () => S,
  extraArgument: E
) => R;

export type TRPGDispatch = ThunkDispatch<any, any, AnyAction>;

// 用于组件声明
export interface TRPGDispatchProp {
  dispatch: TRPGDispatch;
}

// 用于action声明
export type TRPGAction =
  | ThunkAction<any, TRPGState, any, AnyAction>
  | AnyAction;

export type TRPGMiddleware = Middleware<{}, TRPGState, TRPGDispatch>;

interface TRPGThunkApiConfig {
  state: TRPGState;
  dispatch: TRPGDispatch;
  extra?: unknown;
  rejectValue?: unknown;
}

export function createTRPGAsyncThunk<ThunkArg, Returned = void>(
  typePrefix: string,
  payloadCreator: AsyncThunkPayloadCreator<
    Returned,
    ThunkArg,
    TRPGThunkApiConfig
  >,
  options?: any
): AsyncThunk<Returned, ThunkArg, TRPGThunkApiConfig> {
  return createAsyncThunk(typePrefix, payloadCreator, options);
}
