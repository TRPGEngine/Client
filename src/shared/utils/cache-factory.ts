import { getStorage } from '@shared/manager/storage';
import _isNil from 'lodash/isNil';

export enum CachePolicy {
  /**
   * 存储在内存中
   */
  Temporary,

  /**
   * 存储在rnStorage中
   * 过期时间为rnStorage的过期时间
   * NOTE: 在使用持久化存储时必须自行实现hash
   * 可以基于buildCacheHashPrefix来快速实现
   */
  Persistent,
}

/**
 * 序列化参数
 */
function stringifyArgs(arg: any): string {
  switch (typeof arg) {
    case 'string':
      return arg;
    case 'number':
      return String(arg);
    default:
      return JSON.stringify(arg);
  }
}

/**
 * 拼接参数
 */
function joinArgs(args: any[]): string {
  return args.map(stringifyArgs).join('$');
}

function defaultHash(...args: any[]): string {
  return `cache:factory$${joinArgs(args)}`;
}

/**
 * 快速构建缓存的hash
 * @param prefix 前缀
 */
export function buildCacheHashFnPrefix(
  prefix: string
): (...args: any[]) => string {
  return (...args: any[]) => `cache:factory$${prefix}:${joinArgs(args)}`;
}

export function buildCacheFactory<T>(
  policy: CachePolicy,
  fn: (...args: any[]) => any,
  hashFn: (...args: any[]) => string = defaultHash
): (...args: any[]) => Promise<T> {
  let cacheManager: CacheManager;
  if (policy === CachePolicy.Temporary) {
    cacheManager = new TemporaryCache();
  } else if (policy === CachePolicy.Persistent) {
    cacheManager = new PersistentCache();
  }

  return async (...args: any[]) => {
    const hash = hashFn(...args);

    const cache = await cacheManager.get(hash);
    if (!_isNil(cache)) {
      return cache;
    }

    const ret = await fn(...args);
    await cacheManager.set(hash, ret);

    return ret;
  };
}

interface CacheManager {
  get(hash: string): Promise<any>;

  set(hash: string, value: any): Promise<void>;
}

class TemporaryCache implements CacheManager {
  temporaryCacheMap = {};

  async get(hash: string) {
    return this.temporaryCacheMap[hash];
  }
  async set(hash: string, value: any) {
    this.temporaryCacheMap[hash] = value;
  }
}

class PersistentCache implements CacheManager {
  async get(hash: string) {
    return await getStorage().get(hash);
  }
  async set(hash: string, value: any) {
    getStorage().set(hash, value);
  }
}
