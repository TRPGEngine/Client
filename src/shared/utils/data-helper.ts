import type { GroupActorType } from '@src/shared/redux/types/group';
import _get from 'lodash/get';
import _isNil from 'lodash/isNil';
import type { PlayerUser } from '@shared/model/player';
import type { MsgPayload } from '@redux/types/chat';

/**
 * 数据相关帮助函数
 */

/**
 * 获取团人物卡字段
 * @param groupActor 团人物卡
 * @param field 字段名
 */
export const getGroupActorField = (
  groupActor: GroupActorType,
  field: 'avatar' | 'name' | 'desc'
): string => {
  return groupActor[field] || _get(groupActor, ['actor', field]);
};
export const getGroupActorName = (groupActor: GroupActorType) =>
  getGroupActorField(groupActor, 'name');

/**
 * 获取团人物卡信息
 * 为在原始团人物信息的基础上复写团人物信息
 * @param groupActor 团人物卡
 */
export const getGroupActorInfo = (groupActor: GroupActorType): {} => {
  return {
    ..._get(groupActor, 'actor.info', {}),
    ...groupActor.actor_info,
  };
};

/**
 * 获取团人物卡的模板UUID
 * @param groupActor 团人物卡
 */
export const getGroupActorTemplateUUID = (
  groupActor: GroupActorType
): string => {
  return (
    groupActor.actor_template_uuid ?? _get(groupActor, 'actor.template_uuid')
  );
};

/**
 * 获取用户名
 * @param userInfo 用户信息
 */
export function getUserName(
  userInfo: Partial<PlayerUser> | null | undefined
): string {
  if (_isNil(userInfo)) {
    return '';
  }

  return userInfo?.nickname || userInfo?.username || '';
}

export function buildMsgPayload(payload: Partial<MsgPayload>): MsgPayload {
  return {
    type: 'normal',
    uuid: '',
    sender_uuid: '',
    message: '',
    is_public: false,
    is_group: false,
    date: new Date().toISOString(),
    revoke: false,
    ...payload,
  };
}
