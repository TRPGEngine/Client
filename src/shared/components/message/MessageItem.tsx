import React from 'react';
import { TMemo } from '@shared/components/TMemo';
import MessageHandler from './MessageHandler';
import { useCurrentUserInfo } from '@redux/hooks/user';
import { useCachedUserInfo } from '@shared/hooks/useCache';
import type { MsgPayload } from '@redux/types/chat';
import { getUserName } from '@shared/utils/data-helper';

interface Props {
  data: MsgPayload;
  emphasizeTime: boolean;
  omitSenderInfo?: boolean;
}
export const MessageItem: React.FC<Props> = TMemo((props) => {
  const { data, emphasizeTime, omitSenderInfo } = props;
  const selfInfo = useCurrentUserInfo();
  const senderUUID = data.sender_uuid;
  const isMe = selfInfo.uuid === senderUUID;
  const senderInfo = useCachedUserInfo(senderUUID);
  const name = getUserName(senderInfo);
  const avatar = senderInfo.avatar;

  return (
    <MessageHandler
      type={data.type}
      me={isMe}
      name={name}
      avatar={avatar ?? ''}
      emphasizeTime={emphasizeTime}
      omitSenderInfo={omitSenderInfo}
      info={data}
    />
  );
});
MessageItem.displayName = 'MessageItem';
