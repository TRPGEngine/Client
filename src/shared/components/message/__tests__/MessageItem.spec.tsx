/* eslint-disable @typescript-eslint/tslint/config */
/* eslint-disable react/jsx-key */

import { buildMsgPayload } from '@shared/utils/data-helper';
import { TestFullProvider } from '@test/utils';
import { render } from '@testing-library/react';
import React from 'react';
import { MessageItem } from '../MessageItem';
import '@web/components/messageTypes/__all__';
import mockdate from 'mockdate';

describe('MessageItem', () => {
  const name = 'Tester';
  const sender_uuid = 'uuid#0';
  const message = 'Any Message';

  mockdate.set('2021-03-30 12:00:00');

  afterAll(() => {
    mockdate.reset();
  });

  test.each([
    [
      <MessageItem
        data={buildMsgPayload({
          sender_uuid,
          message,
          data: {
            name,
          },
        })}
        emphasizeTime={true}
      />,
    ],
    [
      <MessageItem
        data={buildMsgPayload({
          sender_uuid,
          message,
        })}
        emphasizeTime={false}
      />,
    ],
    [
      <MessageItem
        data={buildMsgPayload({
          sender_uuid: 'trpgbot',
          message,
          data: {
            name,
          },
        })}
        emphasizeTime={false}
        omitSenderInfo={true}
      />,
    ],
    [
      <MessageItem
        data={buildMsgPayload({
          sender_uuid,
          message,
          data: {
            name,
          },
        })}
        emphasizeTime={false}
        omitSenderInfo={false}
      />,
    ],
  ])('testcase %#', (el) => {
    // console.log('date', moment().valueOf());
    const wrapper = render(el, { wrapper: TestFullProvider });
    expect(wrapper.container).toMatchSnapshot();
  });
});
