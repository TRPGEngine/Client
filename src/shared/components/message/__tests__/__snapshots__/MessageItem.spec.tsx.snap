// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageItem testcase 0 1`] = `
<div>
  <div
    class="sc-hKFxyN iMqKQT normal bubble"
  >
    <div
      class="sc-gXfVKN leCvQc"
    >
      <span>
        中午 12:00
      </span>
    </div>
    <div
      class="profile"
    >
      <span
        class="name"
      >
        Tester
      </span>
      <span
        class="time"
      >
        12:00:00
      </span>
    </div>
    <div
      class="content"
    >
      <div
        class="avatar"
      >
        <span
          class="ant-avatar ant-avatar-circle"
          name="Tester"
          style="width: 38px; height: 38px; line-height: 38px; font-size: 15.200000000000001px; cursor: inherit; user-select: none; background-color: rgb(211, 84, 0);"
        >
          <span
            class="ant-avatar-string"
            style="line-height: 38px; transform: scale(1) translateX(-50%);"
          >
            T
          </span>
        </span>
      </div>
      <div
        class="body"
      >
        <pre
          class="sc-bdnxRM gfkJP bubble"
        >
          <pre
            style="display: inline;"
          >
            Any Message
          </pre>
        </pre>
        <div
          class="operation"
        >
          <i
            class="iconfont"
          >
            
          </i>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageItem testcase 1 1`] = `
<div>
  <div
    class="sc-hKFxyN iMqKQT normal bubble"
  >
    <div
      class="profile"
    >
      <span
        class="name"
      />
      <span
        class="time"
      >
        12:00:00
      </span>
    </div>
    <div
      class="content"
    >
      <div
        class="avatar"
      >
        <span
          class="ant-avatar ant-avatar-circle"
          name=""
          style="width: 38px; height: 38px; line-height: 38px; font-size: 15.200000000000001px; cursor: inherit; user-select: none; background-color: rgb(255, 255, 255);"
        >
          <span
            class="ant-avatar-string"
            style="line-height: 38px; transform: scale(1) translateX(-50%);"
          />
        </span>
      </div>
      <div
        class="body"
      >
        <pre
          class="sc-bdnxRM gfkJP bubble"
        >
          <pre
            style="display: inline;"
          >
            Any Message
          </pre>
        </pre>
        <div
          class="operation"
        >
          <i
            class="iconfont"
          >
            
          </i>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageItem testcase 2 1`] = `
<div>
  <div
    class="sc-hKFxyN iMqKQT omitSenderInfo normal bubble"
  >
    <div
      class="content"
    >
      <div
        class="avatar"
      />
      <div
        class="body"
      >
        <pre
          class="sc-bdnxRM gfkJP bubble"
        >
          <pre
            style="display: inline;"
          >
            Any Message
          </pre>
        </pre>
        <div
          class="operation"
        >
          <i
            class="iconfont"
          >
            
          </i>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageItem testcase 3 1`] = `
<div>
  <div
    class="sc-hKFxyN iMqKQT normal bubble"
  >
    <div
      class="profile"
    >
      <span
        class="name"
      >
        Tester
      </span>
      <span
        class="time"
      >
        12:00:00
      </span>
    </div>
    <div
      class="content"
    >
      <div
        class="avatar"
      >
        <span
          class="ant-avatar ant-avatar-circle"
          name="Tester"
          style="width: 38px; height: 38px; line-height: 38px; font-size: 15.200000000000001px; cursor: inherit; user-select: none; background-color: rgb(211, 84, 0);"
        >
          <span
            class="ant-avatar-string"
            style="line-height: 38px; transform: scale(1) translateX(-50%);"
          >
            T
          </span>
        </span>
      </div>
      <div
        class="body"
      >
        <pre
          class="sc-bdnxRM gfkJP bubble"
        >
          <pre
            style="display: inline;"
          >
            Any Message
          </pre>
        </pre>
        <div
          class="operation"
        >
          <i
            class="iconfont"
          >
            
          </i>
        </div>
      </div>
    </div>
  </div>
</div>
`;
